### ISSUE TITLE
Heartbeat issue without Topic

### ISSUE THREAD

**Issue by andysCaplin** [hazelcast/hazelcast-cpp-client#896]
Hi @yemreinci @ihsandemir - is it possible that the issue https://github.com/hazelcast/hazelcast-cpp-client/issues/883 happens without using Topics ?

We've recently had heartbeat issues at some of our servers for clients that are just updating a map and posting messages to queues.

```
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.cached.thread-1] c.h.internal.nio.tcp.TcpIpConnection     : [10.42.3.180]:5701 [soaktest] [4.0.3] Connection[id=8, /10.42.3.180:5701->/10.42.4.251:42440, qualifier=null, endpoint=[10.42.4.251]:42440, alive=false, connectionType=CPP] closed. Reason: Client heartbeat is timed out, closing connection to Connection[id=8, /10.42.3.180:5701->/10.42.4.251:42440, qualifier=null, endpoint=[10.42.4.251]:42440, alive=true, connectionType=CPP]. Now: 2021-07-02 15:07:04.735. LastTimePacketReceived: 2021-07-02 15:06:31.203
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.cached.thread-1] c.h.internal.nio.tcp.TcpIpConnection     : [10.42.3.180]:5701 [soaktest] [4.0.3] Connection[id=25, /10.42.3.180:5701->/10.42.4.6:39452, qualifier=null, endpoint=[10.42.4.6]:39452, alive=false, connectionType=CPP] closed. Reason: Client heartbeat is timed out, closing connection to Connection[id=25, /10.42.3.180:5701->/10.42.4.6:39452, qualifier=null, endpoint=[10.42.4.6]:39452, alive=true, connectionType=CPP]. Now: 2021-07-02 15:07:04.736. LastTimePacketReceived: 2021-07-02 15:06:31.861
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.cached.thread-1] c.h.internal.nio.tcp.TcpIpConnection     : [10.42.3.180]:5701 [soaktest] [4.0.3] Connection[id=95, /10.42.3.180:5701->/10.42.4.7:50732, qualifier=null, endpoint=[10.42.4.7]:50732, alive=false, connectionType=CPP] closed. Reason: Client heartbeat is timed out, closing connection to Connection[id=95, /10.42.3.180:5701->/10.42.4.7:50732, qualifier=null, endpoint=[10.42.4.7]:50732, alive=true, connectionType=CPP]. Now: 2021-07-02 15:07:04.736. LastTimePacketReceived: 2021-07-02 15:06:24.793
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.cached.thread-1] c.h.internal.nio.tcp.TcpIpConnection     : [10.42.3.180]:5701 [soaktest] [4.0.3] Connection[id=96, /10.42.3.180:5701->/10.42.3.192:41886, qualifier=null, endpoint=[10.42.3.192]:41886, alive=false, connectionType=CPP] closed. Reason: Client heartbeat is timed out, closing connection to Connection[id=96, /10.42.3.180:5701->/10.42.3.192:41886, qualifier=null, endpoint=[10.42.3.192]:41886, alive=true, connectionType=CPP]. Now: 2021-07-02 15:07:04.736. LastTimePacketReceived: 2021-07-02 15:06:25.628
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.cached.thread-1] c.h.internal.nio.tcp.TcpIpConnection     : [10.42.3.180]:5701 [soaktest] [4.0.3] Connection[id=97, /10.42.3.180:5701->/10.42.3.193:40126, qualifier=null, endpoint=[10.42.3.193]:40126, alive=false, connectionType=CPP] closed. Reason: Client heartbeat is timed out, closing connection to Connection[id=97, /10.42.3.180:5701->/10.42.3.193:40126, qualifier=null, endpoint=[10.42.3.193]:40126, alive=true, connectionType=CPP]. Now: 2021-07-02 15:07:04.736. LastTimePacketReceived: 2021-07-02 15:06:25.683
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.event-3] c.h.client.impl.ClientEndpointManager    : [10.42.3.180]:5701 [soaktest] [4.0.3] Destroying ClientEndpoint{connection=Connection[id=95, /10.42.3.180:5701->/10.42.4.7:50732, qualifier=null, endpoint=[10.42.4.7]:50732, alive=false, connectionType=CPP], clientUuid='d62eed17-bd74-4af2-bf24-cdc35e16ec19, authenticated=true, clientVersion=4.0.1, creationTime=1625238374848, latest clientAttributes=null}
2021-07-02 15:07:04.736  INFO 15 --- [hz.naughty_pascal.event-3] c.c.d.a.ClientCleanupService             : Client <d62eed17-bd74-4af2-bf24-cdc35e16ec19> has disconnected: ClientEvent{uuid='d62eed17-bd74-4af2-bf24-cdc35e16ec19', eventType=DISCONNECTED, address=/10.42.4.7:50732, clientType=CPP, name='hz.client_1', attributes=[]}
```

**Comment by andysCaplin** [comment]
I added the suggested work-around in issue #883 - a manual heartbeat and got these errors

```
2021-07-02 18:58:48.629 ERROR 15 --- [hz.frosty_grothendieck.partition-operation.thread-1] c.h.map.impl.operation.GetOperation      : [10.42.3.202]:5701 [soaktest] [4.0.3] NormalResponse already responseReceived for callback: Invocation{op=com.hazelcast.map.impl.operation.GetOperation{serviceName='hz:impl:mapService', identityHash=1973623019, partitionId=111, replicaIndex=0, callId=-10250275, invocationTime=1625252328620 (2021-07-02 18:58:48.620), waitTimeout=-1, callTimeout=60000, name=dummy_map}, tryCount=250, tryPauseMillis=500, invokeCount=26, callTimeoutMillis=60000, firstInvocationTimeMs=1625252322603, firstInvocationTime='2021-07-02 18:58:42.603', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.42.3.202]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=null}, current-response: null
com.hazelcast.spi.exception.ResponseAlreadySentException: NormalResponse already responseReceived for callback: Invocation{op=com.hazelcast.map.impl.operation.GetOperation{serviceName='hz:impl:mapService', identityHash=1973623019, partitionId=111, replicaIndex=0, callId=-10250275, invocationTime=1625252328620 (2021-07-02 18:58:48.620), waitTimeout=-1, callTimeout=60000, name=dummy_map}, tryCount=250, tryPauseMillis=500, invokeCount=26, callTimeoutMillis=60000, firstInvocationTimeMs=1625252322603, firstInvocationTime='2021-07-02 18:58:42.603', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.42.3.202]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=null}, current-response: null
2021-07-02 18:58:48.629 ERROR 15 --- [hz.frosty_grothendieck.partition-operation.thread-1] c.h.map.impl.operation.GetOperation      : [10.42.3.202]:5701 [soaktest] [4.0.3] NormalResponse already responseReceived for callback: Invocation{op=com.hazelcast.map.impl.operation.GetOperation{serviceName='hz:impl:mapService', identityHash=73377016, partitionId=147, replicaIndex=0, callId=-10250276, invocationTime=1625252328620 (2021-07-02 18:58:48.620), waitTimeout=-1, callTimeout=60000, name=dummy_map}, tryCount=250, tryPauseMillis=500, invokeCount=32, callTimeoutMillis=60000, firstInvocationTimeMs=1625252319603, firstInvocationTime='2021-07-02 18:58:39.603', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.42.3.202]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=null}, current-response: null
com.hazelcast.spi.exception.ResponseAlreadySentException: NormalResponse already responseReceived for callback: Invocation{op=com.hazelcast.map.impl.operation.GetOperation{serviceName='hz:impl:mapService', identityHash=73377016, partitionId=147, replicaIndex=0, callId=-10250276, invocationTime=1625252328620 (2021-07-02 18:58:48.620), waitTimeout=-1, callTimeout=60000, name=dummy_map}, tryCount=250, tryPauseMillis=500, invokeCount=32, callTimeoutMillis=60000, firstInvocationTimeMs=1625252319603, firstInvocationTime='2021-07-02 18:58:39.603', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.42.3.202]:5701, pendingResponse={VOID}, backupsAcksExpected=-1, backupsAcksReceived=0, connection=null}, current-response: null
```

**Comment by yemreinci** [comment]
> is it possible that the issue #883 happens without using Topics ?

Yes, the issue wasn't specific to Topic. It happens when a client receives frequent events from the cluster (queue events, map events, topic events etc.) but doesn't send requests for longer than the heartbeat timeout duration. The heartbeat sending logic had an error causing it to not send heartbeats for such cases. They will all be covered by [the fix](https://github.com/hazelcast/hazelcast-cpp-client/pull/891). 

About the second error, one thing caught my attention in the shared log line: `callId=-10250275`. Call id is negative. @ihsandemir, is that normal?

**Comment by andysCaplin** [comment]
Hi. I haven't seen that callId parameter in logs before before. What does it signify ?

When I implemented the dummy heart beat and tested locally I didn't get any error logs - but I only have one server when I test locally.
The error logs are from our soak system where we're testing bringing clients and servers up and taking them down.
The dummy heartbeat definitely fixes the HB disconnects because we're not getting them any more.
But the main issue we have seems to relate to how servers and clients interact especially when we have multiple servers and we take a server down.
The long delay we get for the queue response only seems to happen when we remove a server.

Our usage of Hazelcast is fairly simple
- We have Java servers and C++ clients.
- We have 1 global map and 2 queues pairs per client ( the Topic usage is for a new feature which is not in test yet )
- Each client adds itself to the map when it starts and is removed when it stops or disconnects
- 1 queue pair is only used for client startup - to control how many clients are allowed in the system - so it's not busy. 1 message posted by client when it starts and one response posted by server.
- The other queue pair is busy especially in the soak test, multiple messages posted by clients and responses posted by servers.

The second queue pair is the one that has the issues. We see clients timing out on expected responses ( 30 seconds is allowed by our app ) and the expected responses received minutes later.
But again - we only see this when we stop the servers.
Is there anything in the HB issue that might cause a symptom like this ?

**Comment by ihsandemir** [comment]
> > is it possible that the issue #883 happens without using Topics ?
> 
> Yes, the issue wasn't specific to Topic. It happens when a client receives frequent events from the cluster (queue events, map events, topic events etc.) but doesn't send requests for longer than the heartbeat timeout duration. The heartbeat sending logic had an error causing it to not send heartbeats for such cases. They will all be covered by [the fix](https://github.com/hazelcast/hazelcast-cpp-client/pull/891).
> 
> About the second error, one thing caught my attention in the shared log line: `callId=-10250275`. Call id is negative. @ihsandemir, is that normal?

Hello, I would not expect the call id become negative unless there were too many connections (more than 2 billion opened and closed so that the int32_t overflowed). Our 8 byte call id that we send to the server is composed of the first bytes being the connection id and the next 4-bytes is an incremental id for the message. Hence, we combine the two and form a call id. Here is [the code for this](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h#L102): 
```
                                struct correlation_id {
                                    int32_t connnection_id;
                                    int32_t call_id;
                                };
                                union {
                                    int64_t id;
                                    correlation_id composed_id;
                                } c_id_union;
                                c_id_union.composed_id = {connection->get_connection_id(), call_id};
                                message_call_id = c_id_union.id;
                                message->set_correlation_id(c_id_union.id);
```

**Comment by ihsandemir** [comment]
> Hi. I haven't seen that callId parameter in logs before before. What does it signify ?

Hello @andysCaplin callId is the unique 8 byte number that is used to track an invocation. The client always generates this number. It is only unique per connection to the server. This id is written in the message sent to the server as well as the response received from the server, and hence we always use this id to track the response to the original request and the events.

**Comment by andysCaplin** [comment]
The only time I've seen that error and that callId was when I added the heartbeat work-around, i.e.

```
cli->get_map("dummy_map").get()->get<int, int>(rand());
```

I added a timed event to call that every second. The clients that were accessing that map number no more than 20 or 30 max and they were running for hours, occasionally being stopped and re-started. But I don't think it was long enough to get that many calls.

Also, what does the server error - ResponseAlreadySentException - mean ? Surely a client can get a map entry as often as it wants and it doesn't matter if the response was already sent ?

**Comment by ihsandemir** [comment]
@andysCaplin Let's try to reproduce the problem and investigate what is happening. 

`ResponseAlreadySentException` may not be important since it is a map.get call which is a retryable operation (and it is idempotent, i.e. produces the same result even if multiple invocations).

When you said that the response comes in minutes when the server restarts, we need to investigate this case as well. I would not normally expect such delays. There are multiple things occur when a server is restarted: When it shuts down, the partitions re-arrangement among the existing servers can be triggered, we call this `partition migration` so that a new balance is established. If the data is big, this may take a while but the requests should be getting responses pretty quick. What is the data size on the bigger map that you mention which is slow? If we can generate the case, we can see what the client is waiting on for that time.

**Comment by andysCaplin** [comment]
Thanks for getting back on this.

Re the ResponseAlreadySentException - we can live with that because we'll only have the work-around in until the new release anyway.

Re the queue slow response issue, I'm off today but I'll be doing another testing cycle with the servers tomorrow and we'll focus on that issue to try to characterise it further. 
As far as we know so far from the logging, the server seems to send the response straight away, which suggests the delay is in the C++ client. But is is possible that we are just logging that the response is sent and it sits at the server before being sent.

**Comment by andysCaplin** [comment]
Some more info. I was running my code under valgrind ( this is the Topic feature where I found the original heartbeat issue ) and the dummy map work-around gave the issue shown below in valgrind.
Again, the dummy heartbeat does stop the heartbeat loss and disconnection.

```
==10916== Thread 3 trans-Discovery:
==10916== Invalid read of size 4
==10916==    at 0xBFE4B0: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_add_ref_lock() (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xBF01E4: hazelcast::client::spi::impl::ClientInvocation::ClientInvocation(hazelcast::client::spi::ClientContext&, std::shared_ptr<hazelcast::client::protocol::ClientMessage>&&, std::string const&, int, std::shared_ptr<hazelcast::client::connection::Connection> const&, boost::uuids::uuid) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xBF071B: hazelcast::client::spi::impl::ClientInvocation::create(hazelcast::client::spi::ClientContext&, std::shared_ptr<hazelcast::client::protocol::ClientMessage>&&, std::string const&, std::shared_ptr<hazelcast::client::connection::Connection> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xBF0822: hazelcast::client::spi::impl::ClientInvocation::create(hazelcast::client::spi::ClientContext&, hazelcast::client::protocol::ClientMessage&, std::string const&, std::shared_ptr<hazelcast::client::connection::Connection> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xBF7D89: hazelcast::client::spi::ProxyManager::initialize(std::shared_ptr<hazelcast::client::spi::ClientProxy> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xAAF0DC: boost::shared_future<std::shared_ptr<hazelcast::client::imap> > hazelcast::client::spi::ProxyManager::get_or_create_proxy<hazelcast::client::imap>(std::string const&, std::string const&) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0xA91C81: boost::shared_future<std::shared_ptr<hazelcast::client::imap> > hazelcast::client::impl::hazelcast_client_instance_impl::get_distributed_object<hazelcast::client::imap>(std::string const&) (in /home/andys/clones/cdev/rtas/build/transformer)
==10916==    by 0x9CF8DD: hazelcast::client::hazelcast_client::get_map(std::string const&) (hazelcast_client.h:100)
==10916==    by 0x9B9D79: dummy_heartbeat (discovery.cpp:144)
```

**Comment by ihsandemir** [comment]
> Some more info. I was running my code under valgrind ( this is the Topic feature where I found the original heartbeat issue ) and the dummy map work-around gave the issue shown below in valgrind.
> Again, the dummy heartbeat does stop the heartbeat loss and disconnection.
> 
> The exit() call show below is in our signal handler. In this case SIGINT is being handled and by the time we call exit(), the handler has called the Hazelcast shutdown routine.
> 
> ```
> ==10916== Thread 3 trans-Discovery:
> ==10916== Invalid read of size 4
> ==10916==    at 0xBFE4B0: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_add_ref_lock() (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xBF01E4: hazelcast::client::spi::impl::ClientInvocation::ClientInvocation(hazelcast::client::spi::ClientContext&, std::shared_ptr<hazelcast::client::protocol::ClientMessage>&&, std::string const&, int, std::shared_ptr<hazelcast::client::connection::Connection> const&, boost::uuids::uuid) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xBF071B: hazelcast::client::spi::impl::ClientInvocation::create(hazelcast::client::spi::ClientContext&, std::shared_ptr<hazelcast::client::protocol::ClientMessage>&&, std::string const&, std::shared_ptr<hazelcast::client::connection::Connection> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xBF0822: hazelcast::client::spi::impl::ClientInvocation::create(hazelcast::client::spi::ClientContext&, hazelcast::client::protocol::ClientMessage&, std::string const&, std::shared_ptr<hazelcast::client::connection::Connection> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xBF7D89: hazelcast::client::spi::ProxyManager::initialize(std::shared_ptr<hazelcast::client::spi::ClientProxy> const&) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xAAF0DC: boost::shared_future<std::shared_ptr<hazelcast::client::imap> > hazelcast::client::spi::ProxyManager::get_or_create_proxy<hazelcast::client::imap>(std::string const&, std::string const&) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0xA91C81: boost::shared_future<std::shared_ptr<hazelcast::client::imap> > hazelcast::client::impl::hazelcast_client_instance_impl::get_distributed_object<hazelcast::client::imap>(std::string const&) (in /home/andys/clones/cdev/rtas/build/transformer)
> ==10916==    by 0x9CF8DD: hazelcast::client::hazelcast_client::get_map(std::string const&) (hazelcast_client.h:100)
> ==10916==    by 0x9B9D79: dummy_heartbeat (discovery.cpp:144)
> 
> 
> ==10916==  Address 0x10224330 is 7,120 bytes inside a block of size 12,928 free'd
> ==10916==    at 0x4C28DAC: operator delete(void*) (vg_replace_malloc.c:584)
> ==10916==    by 0x9E1A47: std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release() (shared_ptr_base.h:155)
> ==10916==    by 0x5C31A01: exit (in /lib64/libc-2.12.so)
> ==10916==    by 0x99671A: pipe_signal_callback (pipe.c:93)
> ==10916==    by 0x964E77: caplinutil_signalsafe_read_message (signalsafe.c:80)
> ```

Are you stopping the `dummy_heartbeat` thread before shutting  down the client before exiting?

**Comment by andysCaplin** [comment]
Yeah - sorry. You're right. I wasn't.

**Comment by andysCaplin** [comment]
We're updating the code to use TTL as suggested and I'll be doing this every second
```
peerMap->set(uuid, caplinPeer, c_peer_ttl);
```

Will that fix the heartbeat issue or does it specifically need a map get as was shown in the work around example ?

**Comment by andysCaplin** [comment]
I can see that the other thing in the example was a call to rand() - was a changing key also requested ?

I've now found 2 issues - the second one may relate to the queue delay issue - #896.
I removed the dummy HB and put a map get into the TTL timed event so that every second we do a get on the peer map entry as well as setting the TTL.
This timed event is deleted when we disconnect from the server but when we do disconnect the Library looks like it is still trying to get an entry from the map every second

```
This is our timed event resetting TTL and reading the peer entry
2021/07/06-14:46:05.560302 +0100: (rttpd-Discovery): FINEST: Discovery: Peer - Reset TTL for <caplin.peers><3b5a0f78-8848-4ef0-b1e8-36c7b6bd1738>
Entry is read and deserialised
2021/07/06-14:46:05.591423 +0100: (1844): DEBUG: Discovery: Peer - Deserialised peer <winlib>

Next TTL
2021/07/06-14:46:06.558182 +0100: (rttpd-Discovery): FINEST: Discovery: Peer - Reset TTL for <caplin.peers><3b5a0f78-8848-4ef0-b1e8-36c7b6bd1738>

This is the disconnect ( happens before the read is complete ? )
2021/07/06-14:46:06.590392 +0100: (16932): INFO: Discovery: Library - <hz.client_1><dev> ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=2021-07-06 14:46:06.000, connected server version=4.0.3} closed. Reason: Socket read error. system:1236 for ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=never, connected server version=4.0.3}
2021/07/06-14:46:06.590633 +0100: (16932): INFO: Discovery: Library - <hz.client_1><dev> Removed connection to endpoint: Address[172.20.0.239:5701], connection: ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-2, closedTime=2021-07-06 14:46:06.0-1, connected server version=4.0.3}
2021/07/06-14:46:06.590680 +0100: (16932): INFO: Discovery: Library - <hz.client_1><dev> LifecycleService::LifecycleEvent CLIENT_DISCONNECTED
2021/07/06-14:46:06.590832 +0100: (rttpd-Discovery): NOTIFY: Discovery: Server disconnected
2021/07/06-14:46:06.590953 +0100: (rttpd-Discovery): NOTIFY: Discovery: Instance <3b5a0f78-8848-4ef0-b1e8-36c7b6bd1738> of <rttpd> is no longer licensed
2021/07/06-14:46:06.591051 +0100: (17996): INFO: Discovery: Library - <hz.client_1><dev> Trying to connect to Address[172.20.0.239:5701]. Translated address:Address[172.20.0.239:5701].

Here we delete the timed event so we're not getting from the map any more. I don't get any more "Reset TTL" logs so I know the timed event is gone.
2021/07/06-14:46:06.613661 +0100: (rttpd-Discovery): FINEST: Discovery: Peer - Stop TTL for <caplin.peers><3b5a0f78-8848-4ef0-b1e8-36c7b6bd1738>

But now on a Hazelcast thread I start getting lots of these logs for about 2 seconds - all referring to a get from the map I'm not getting from any more.

2021/07/06-14:46:06.614035 +0100: (16932): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=22, operation=client.addclusterviewlistener, isRetryable=0, correlationId = 34058910268653569, messageType = 0x300, flags = 0xe100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = , target = connection ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-25, closedTime=2021-07-06 14:46:06.-24, connected server version=4.0.3}, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-25, closedTime=2021-07-06 14:46:06.-24, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 1, exception : target_disconnected {Socket read error. system:1236 for ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=never, connected server version=4.0.3}. Error code:52, Details:.} at Connection::close.
2021/07/06-14:46:06.619832 +0100: (16932): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=22, operation=client.localbackuplistener, isRetryable=0, correlationId = 34058914563620865, messageType = 0xf00, flags = 0xe100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = , target = connection ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-31, closedTime=2021-07-06 14:46:06.-30, connected server version=4.0.3}, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-31, closedTime=2021-07-06 14:46:06.-30, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 1, exception : target_disconnected {Socket read error. system:1236 for ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=never, connected server version=4.0.3}. Error code:52, Details:.} at Connection::close.

Then we get several of these

2021/07/06-14:46:06.648069 +0100: (16884): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=102, operation=map.get, isRetryable=1, correlationId = 136, messageType = 0x10200, flags = 0xc100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = caplin.peers, target = partition 233, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-59, closedTime=2021-07-06 14:46:06.-57, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 2, exception : io {No connection found to cluster.. Error code:22, Details:.} at ClientConnectionManagerImpl::check_invocation_allowed.

Which eventually start coming out every second

```

When I reconnect the network we connect successfully and the server sends us a response message on a queue which we never get - probably because of this log

```
2021/07/06-14:51:37.077378 +0100: (16624): FINEST: Discovery: Library - <hz.client_1><dev>  We will not apply the response, since response state version is old . Response is from ClientConnection{alive=1, connectionId=9, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:51:37.000, closedTime=never, connected server version=4.0.3}. Current connection ClientConnection{alive=1, connectionId=9, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:51:37.000, closedTime=never, connected server version=4.0.3}, response state version:272. Current state version: 272
```

Do clients need to tear down and recreate queue listeners when they disconnect/reconnect to the server ?

**Comment by andysCaplin** [comment]
All the issues above go away if I don't get from the map periodically.

**Comment by ihsandemir** [comment]
> We're updating the code to use TTL as suggested and I'll be doing this every second
> 
> ```
> peerMap->set(uuid, caplinPeer, c_peer_ttl);
> ```
> 
> Will that fix the heartbeat issue or does it specifically need a map get as was shown in the work around example ?

No, you do not need to do `get_map` for each call. It will return the cached proxy anyway. Just use it to do an invocation. Doing a map->set every second is similar to heartbeat but it will update only the connections where to the server where the key exists (based on its associated partition). Hence, that is why I asked you to do a `map->get` on a random key. If the key you are using is hitting all the connected servers, then you do not need an extra dummy heartbeat thread.

**Comment by ihsandemir** [comment]
> I can see that the other thing in the example was a call to rand() - was a changing key also requested ?

Yes, changing the key to make sure that we create traffic on all the connections if there are multiple servers.

> But now on a Hazelcast thread I start getting lots of these logs for about 2 seconds - all referring to a get from the map I'm not getting from any more.
> 
> ```
> 2021/07/06-14:46:06.614035 +0100: (16932): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=22, operation=client.addclusterviewlistener, isRetryable=0, correlationId = 34058910268653569, messageType = 0x300, flags = 0xe100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = , target = connection ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-25, closedTime=2021-07-06 14:46:06.-24, connected server version=4.0.3}, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-25, closedTime=2021-07-06 14:46:06.-24, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 1, exception : target_disconnected {Socket read error. system:1236 for ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=never, connected server version=4.0.3}. Error code:52, Details:.} at Connection::close.
> 2021/07/06-14:46:06.619832 +0100: (16932): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=22, operation=client.localbackuplistener, isRetryable=0, correlationId = 34058914563620865, messageType = 0xf00, flags = 0xe100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = , target = connection ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-31, closedTime=2021-07-06 14:46:06.-30, connected server version=4.0.3}, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-31, closedTime=2021-07-06 14:46:06.-30, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 1, exception : target_disconnected {Socket read error. system:1236 for ClientConnection{alive=1, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.0-1, closedTime=never, connected server version=4.0.3}. Error code:52, Details:.} at Connection::close.
> ```
> 

These can be OK. There was an internal invocation to the server, one is `operation=client.addclusterviewlistener` and the other is `operation=client.localbackuplistener`, which failed when the connection is lost. This should not be a problem at all but just logs.

> 
> ```
> 2021/07/06-14:46:06.648069 +0100: (16884): FINEST: Discovery: Library - <hz.client_1><dev> Invocation got an exception ClientInvocation{requestMessage = ClientMessage{length=102, operation=map.get, isRetryable=1, correlationId = 136, messageType = 0x10200, flags = 0xc100, is backup aware = 1, is backup event = 0, isEvent = 0}, is_fragmented = no, objectName = caplin.peers, target = partition 233, sendConnection = ClientConnection{alive=0, connectionId=1, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:46:05.-59, closedTime=2021-07-06 14:46:06.-57, connected server version=4.0.3}, backup_acks_expected_ = -1, backup_acks_received = 0}, invoke count : 2, exception : io {No connection found to cluster.. Error code:22, Details:.} at ClientConnectionManagerImpl::check_invocation_allowed.
> ```
> 

Similar to above reason, it may be repeated in the logs until the invocation retry timeout (default 120secs). 

> 
> When I reconnect the network we connect successfully and the server sends us a response message on a queue which we never get - probably because of this log
> 
> ```
> 2021/07/06-14:51:37.077378 +0100: (16624): FINEST: Discovery: Library - <hz.client_1><dev>  We will not apply the response, since response state version is old . Response is from ClientConnection{alive=1, connectionId=9, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:51:37.000, closedTime=never, connected server version=4.0.3}. Current connection ClientConnection{alive=1, connectionId=9, remoteEndpoint=Address[172.20.0.239:5701], lastReadTime=2021-07-06 14:51:37.000, closedTime=never, connected server version=4.0.3}, response state version:272. Current state version: 272
> ```
> 
> Maybe this is related to our log queue response time ?
> 

This message is also OK, it means that we already have a newer state and we just ignore this older update.

> Do clients need to tear down and recreate queue listeners when they disconnect/reconnect to the server ?

Yes, definitely they re-register the listeners when connected to the servers so that you can still keep getting the events.

**Comment by andysCaplin** [comment]
Re the random key and TTL. I guess that if enough random keys are created they end up spread across all partitions because they are different keys and that means that all the servers get hit.

The key I'm updating with TTL is a single key. I guess it goes on one partition on one server so it won't handle the heartbeat issue when we have multiple servers.

**Comment by andysCaplin** [comment]
> Do clients need to tear down and recreate queue listeners when they disconnect/reconnect to the server ?
> > Yes, the client will add the listeners on reconnection to the servers.

What I meant was - if my application gets a new client ( which has connected ) and creates map/queue listeners, does it need to tear down those listeners when all the server(s) disconnect and then re-create them when one or more of the server(s) connect ?

I thought once we set up the listeners they were handled under the hood by the library.

**Comment by ihsandemir** [comment]
> Re the random key and TTL. I guess that if enough random keys are created they end up spread across all partitions because they are different keys and that means that all the servers get hit.
> 
> The key I'm updating with TTL is a single key. I guess it goes on one partition on one server so it won't handle the heartbeat issue when we have multiple servers.

Exactly. By the way, if you grab the unreleased master code and install the library if it is possible for you, then it already has the fix to the problem and hence, you should not worry about the heartbeat with the master code.

**Comment by ihsandemir** [comment]
> 
> What I meant was - if my application gets a new client ( which has connected ) and creates map/queue listeners, does it need to tear down those listeners when all the server(s) disconnect and then re-create them when one or more of the server(s) connect ?
> 
> I thought once we set up the listeners they were handled under the hood by the library.

Firstly, just lets understand that getting a totally fresh new client is different case. The listeners belong to a client, hence new client means you have to re-register the listeners and configure the new client.

But why would you get a new client each time? You just instantiate one client and reuse it all the time. Normally, you should not need a second client at all. You can do `get_map` call as many times as you want, if it is the same map name, you locally always get the same map pointer, hence if a listener is attached to that map, it always works since it is the same map. When the client reconnects to any member, yes, it re-registers the listeners under the hood for you. Does this clarify the questions?

**Comment by andysCaplin** [comment]
Yes - thanks. I should have been clearer. We never get a new client - we get it initially and reuse. 

I assumed that what you said above was how it worked. 

I just wanted to be certain because we definitely seem to get issues with client listeners not getting responses sent by the server after a server restarts.

I did a gstack when this happens and none of the threads are stuck.

I'll be testing this again to characterise the issue further.

**Comment by andysCaplin** [comment]
@ihsandemir we're creating a simple C++ client and Java server to isolate the queue response delay issue. When we get them finished and have reproduced the issue with them I'll attach them to this issue.

**Comment by ihsandemir** [comment]
> I just wanted to be certain because we definitely seem to get issues with client listeners not getting responses sent by the server after a server restarts.

I definitely like to solve if there is such an issue. The listeners should work seamlessly on cluster reconnects. 

That would be great if you can share a reproducer. Thanks again for the efforts. 

(Note: will be away for two weeks, i will check it when i return.)

**Comment by andysCaplin** [comment]
@ihsandemir We now think that our queue response issues may have been due to re-partitioning occurring at the servers when we scale them up and down. The guy working on our server is trying to address that.

We see a general slowing down including our client TTL saying that it's not being updated at the rate the clients are updating it - this includes Java clients as well as C++ clients.

I think there is a separate issue when we lose connectivity and restore it but I need to do more testing in that area.

**Comment by ihsandemir** [comment]
When the repartitioning occurs, the partitions are migrated incrementally and those migrating partition related operations may be delayed until the migration of the partitions is completed. Queue with a specific name matches to a single partition. Hence, actually in the cluster the queue data stays at a single partition. If that partition is migrating, the queue operations may be delayed. But it should continue when the migration completes.

**Comment by yemreinci** [comment]
> Hello, I would not expect the call id become negative unless there were too many connections (more than 2 billion opened and closed so that the int32_t overflowed). Our 8 byte call id that we send to the server is composed of the first bytes being the connection id and the next 4-bytes is an incremental id for the message. Hence, we combine the two and form a call id. Here is [the code for this](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/hazelcast/include/hazelcast/client/internal/socket/BaseSocket.h#L102):
> 
> ```
>                                 struct correlation_id {
>                                     int32_t connnection_id;
>                                     int32_t call_id;
>                                 };
>                                 union {
>                                     int64_t id;
>                                     correlation_id composed_id;
>                                 } c_id_union;
>                                 c_id_union.composed_id = {connection->get_connection_id(), call_id};
>                                 message_call_id = c_id_union.id;
>                                 message->set_correlation_id(c_id_union.id);
> ```

@ihsandemir I think the call_id being negative implies that `correlation_id::call_id` is negative, not `correlation_id::connection_id` as you suggest (on little endian architectures). The most significant bytes of `composed_id` comes from `correlation_id::call_id` as it comes later in the struct. See the [compiler explorer example](https://godbolt.org/z/1Ez7zTGaE).

So, in this case, I think the number of calls could have overflowed and caused the `ResponseAlreadySentException` as the correlation_id's started to clash. I think it is possible for such an overflow to occur during the lifetime of a client as the call id is stored only in 32 bits. For example, I measured a throughput of ~65K map get calls per second locally. With this rate, the counter would overflow in about ~9 hours.

**Comment by ihsandemir** [comment]
@yemreinci yes, you are right for little endian systems it would be OK to have negative correlation id values.

**Regarding the `ResponseAlreadySentException`**: As you can see from the logs, this is a server side log. See [this line](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/Invocation.java#L222) at the server code. Servers communicate with each other via Invocations and it looks like a response was sent for an invocation and another response is trying to be sent and it is declined and logged. There may be different reasons for this to occur and repartitioning of data may be one of them. In any case, this should not be related to the client side.

**Regarding the call id overflow**: So, we spare 32-bits for the call id per connection, which means around 2 billion outstanding messages for a connection. Normally, the call id is not retained anywhere once the response is received, hence it is not meaningful once the invocation is finished (except for the listener call ids, they stay there for a while). If there are so many outstanding invocations per connection, well that is some real problem, I would normally not expect any such situation, it almost means that there are no responses received and remember that we have a timeout for any invocation a default of 120 seconds, hence, it will be dropped at most after 120seconds. Well, if we want we can play with the call-id versus connection-id ranges. For example, we can move 16 more bites from connection-ids, making the connection id 16-bits and make the call ids 48-bits if desired. But, I really do not think that this is the cause and i can not prove that this is an overflow case.

**Comment by ihsandemir** [comment]
> @ihsandemir We now think that our queue response issues may have been due to re-partitioning occurring at the servers when we scale them up and down. The guy working on our server is trying to address that.
> 
> We see a general slowing down including our client TTL saying that it's not being updated at the rate the clients are updating it - this includes Java clients as well as C++ clients.
> 
> I think there is a separate issue when we lose connectivity and restore it but I need to do more testing in that area.

Hello @andysCaplin Were you able to do more testing, any findings? We currently can not reproduce the problem and just like you mentioned you may be observing temporarily slow down during repartitioning due to the servers being less responsive with data migrations. 

Regarding the listeners, do you have any problems? While the client reconnects to the new servers, you may lose some events but eventually, when all connections are stable, you should always see the events flowing. Please let me know if any further problems.

**Comment by andysCaplin** [comment]
We had some issues that make sense
- Request sent to a server that is being scaled, so it gets the request but doesn't get time to send a response
- Misconfiguration of the servers - where 2 servers on the same host used localhost in the configuration

The 1st one we can work around and the second was a config error.

In our main test system we have 0-3 servers sitting behind a load balancer and our test is periodically scaling the servers and peers up and down.

One thing I've definitely found is that we must avoid polling.

To mitigate the listener issues we saw I added a poller that checked if the queue was empty and read what was on the queue if it wasn't empty. In general this poller never reads anything because the listener always gets in first.

But we found out that if this poll occurs when a server goes down we can sit on the empty check for 5,10, 15 seconds sometimes longer just stuck until things settle down - I assume with repartitioning. Also, the server engineer I'm working with says there's a timeout when a server disconnects - 10 seconds - before the other server(s) decide it's really gone which probably contributes to this time.

The way we use the queues we can handle some lost responses because the queues are managing application events that timeout and retry and we only get these issues when severs disconnect - which we expect to be infrequent.

**Question re listener**

The listener examples I've seen always read a single element from the queue. Is this the recommended way or is it better to check the size of the queue and drain it in the listener event ?

We've also tested our queues over a slow network and we can end up with 100's of entries backed up on the queue. Is it quicker to drain the queue instead of waiting for an event per entry ?

**Comment by ihsandemir** [comment]
> **Question re listener**
> 
> The listener examples I've seen always read a single element from the queue. Is this the recommended way or is it better to check the size of the queue and drain it in the listener event ?
> 

Which example are you referring to? Is [this](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/examples/distributed-collections/blockingqueue/Consumer.cpp#L24) what you are referring? Or are you referring to [listener example](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/examples/distributed-map/entry-listener/main.cpp#L23) for the map? 

> We've also tested our queues over a slow network and we can end up with 100's of entries backed up on the queue. Is it quicker to drain the queue instead of waiting for an event per entry ?

Regarding your question: `drain_to` would be more efficient than `take` or `poll` if there are multiple items in the queue. If there is a single entry probably `take` would be as good or better. Now, event is totally different, events are sent from the server to client one-by-one as efficient as possible as they occur. They are usually pretty fast enough. The only drawback of event is that it is not fully reliable when servers crash or change, some events can be lost. That is why you need an extra check when you use events to make sure that you do not lose events. This extra check can be a to drain (or poll) from the queue in case any events are missed. Usually, people also design incremental ids for events if it is possible, and they can detect which items are being missed and request those items those are missed (of course this means buffering a number of last items).

In your question, if you see 100's of entries backed up, I would just drain them.

**Comment by andysCaplin** [comment]
Hi. Thanks for the information. I was unaware of the poll() and drain_to() API's. When I mentioned poll I meant that I had a timed event that periodically tried a check of the queue being empty and did a take().get() if it wasn't.

The problem with that approach was that the call to is_empty() or take().get() hung if any of the servers disconnected - until the library sorted out partitioning and whatever else it needed to.

In general what I'm finding is that when a server disconnects then any type of read operation (  get, take, isempty etc ) on a queue or map, seems to hang waiting for the library to do the housekeeping associated with server disconnect.
Writes never seem to hang in this instance.

I can't tell if reads within listener events are ok - if they do hang it's harder for me to see because that would be entirely within a Hazelcast thread.

If I use drain_to() or poll() **within an even**t am I guaranteed that they won't hang ? I can see that poll() has a timeout so I assume that is guaranteed not to get stuck but drain_to() doesn't have a timeout.

Finally - we have a similar issue with TTL. We now use TTL to keep our peer entries alive BUT in the C++ API there is no API to set TTL that doesn't also recreate the entry as far as I can see ( in Java there is ).
We want to avoid having the TTL update recreating the entry so I try to get the entry first. If it doesn't exist then I don't set the TTL.
But this initial get has the same problem if a server goes down - it hangs until the library has done various housekeeping actions.

Is there any way around that issue ?

**Comment by ihsandemir** [comment]
Hello, @andysCaplin I think that you have great questions. 

For release purposes, I need to close this issue and can we continue the discussion in our [C++ slack channel](https://hazelcastcommunity.slack.com/archives/C01NC9N8CAE), would that be OK? I will try to answer each question you have in the channel.

**Comment by andysCaplin** [comment]
Yes - sorry about posting here. A lot of what I'm discussing/asking would probably be more appropriate there.

The only question left in relation to the original HB issue is when there is a 4.0 release with the fix. We still have the work-around and it fixes the issue but we'd prefer to release with an official 4.0 with the fix if possible.

Our servers are 4.0 so we need the clients to stay on 4.0 as well.
Also as we're getting close to our release we don't want to risk going away from 4.0.

**Comment by ihsandemir** [comment]
@andysCaplin Sorry, that may be a wrong link. Can you try this: https://slack.hazelcast.com/ join the #cpp-client channel. Here will be more people to share the knowledge.
